### DISCUSSION TITLE
Phase Field: Solve Did NOT Converge!

### DISCUSSION THREAD

**Discussion by wowodejiajia** [idaholab/moose#19234] [Q&A Modules: Phase field] [answered]
Dear MOOSE developers,
I downloaded the input code from the literature. Run it on my own computer.

19 Nonlinear |R| = 3.234805e+06
      0 Linear |R| = 3.234805e+06
      1 Linear |R| = 3.234747e+06
      2 Linear |R| = 3.234728e+06
      3 Linear |R| = 3.234719e+06
      4 Linear |R| = 3.234713e+06
      5 Linear |R| = 3.234709e+06
      6 Linear |R| = 3.234707e+06
      7 Linear |R| = 3.234705e+06
      8 Linear |R| = 3.234703e+06
      9 Linear |R| = 3.234702e+06
     10 Linear |R| = 3.234701e+06
     11 Linear |R| = 3.234700e+06
     12 Linear |R| = 3.234699e+06
     13 Linear |R| = 3.234699e+06
     14 Linear |R| = 3.234698e+06
     15 Linear |R| = 3.234698e+06
     16 Linear |R| = 3.234697e+06
     17 Linear |R| = 3.234697e+06
     18 Linear |R| = 3.234696e+06
     19 Linear |R| = 3.234696e+06
     20 Linear |R| = 3.234695e+06
     21 Linear |R| = 3.234694e+06
     22 Linear |R| = 3.234694e+06
     23 Linear |R| = 3.234694e+06
     24 Linear |R| = 3.234693e+06
     25 Linear |R| = 3.234693e+06
     26 Linear |R| = 3.234693e+06
     27 Linear |R| = 3.234693e+06
     28 Linear |R| = 3.234693e+06
     29 Linear |R| = 3.234693e+06
     30 Linear |R| = 3.234796e+06
     31 Linear |R| = 3.234745e+06
     32 Linear |R| = 3.234725e+06
     33 Linear |R| = 3.234717e+06
     34 Linear |R| = 3.234711e+06
     35 Linear |R| = 3.234708e+06
     36 Linear |R| = 3.234705e+06
     37 Linear |R| = 3.234703e+06
     38 Linear |R| = 3.234701e+06
     39 Linear |R| = 3.234700e+06
     40 Linear |R| = 3.234699e+06
     41 Linear |R| = 3.234698e+06
     42 Linear |R| = 3.234697e+06
     43 Linear |R| = 3.234696e+06
     44 Linear |R| = 3.234696e+06
     45 Linear |R| = 3.234696e+06
     46 Linear |R| = 3.234695e+06
     47 Linear |R| = 3.234695e+06
     48 Linear |R| = 3.234695e+06
     49 Linear |R| = 3.234694e+06
     50 Linear |R| = 3.234694e+06
  Linear solve did not converge due to DIVERGED_ITS iterations 50
20 Nonlinear |R| = 3.234580e+06
Nonlinear solve did not converge due to DIVERGED_MAX_IT iterations 20
 Solve Did NOT Converge!
  Finished Solving                                                                       [179.60 s] [    0 MB]
Aborting as solve did not converge

**Comment by wowodejiajia** [comment]
```
[Mesh]
  # input file.
  # Define mesh. 2-D system, simulation size 200*200.
  type = GeneratedMesh
  dim = 2
  nx = 200
  xmax = 200
  ny = 200
  ymax = 200
[]

[Variables]
  # variables. w: chemical potential, eta: order parameter, pot: applied overpotential.  
  [./w]
  [../]
  [./eta]
  [../]
  [./pot]
  [../]
[]

[Functions]
  # Creating functions for initial conditions.
  [./ic_func_eta]
    type = ParsedFunction
    value = 0.5*(1.0-1.0*tanh((x-20)*2))
  [../]
  [./ic_func_c]
    type = ParsedFunction
    value = 0
  [../]
  [./ic_func_pot]
    type = ParsedFunction
    value = -0.225*(1.0-tanh((x-20)*2))
  [../]
[]

[ICs]
  # Initial conditions.
  [./eta]
    variable = eta
    type = FunctionIC
    function = ic_func_eta
  [../]
  [./w]
    variable = w
    type = FunctionIC
    function = ic_func_c
  [../]
  [./pot]
    variable = pot
    type = FunctionIC
    function = ic_func_pot
  [../]
[]

[BCs]
  # Boundary conditions.
  [./bottom_eta]
    type = NeumannBC
    variable = 'eta'
    boundary = 'bottom'
    value = 0
  [../]
  [./top_eta]
    type = NeumannBC
    variable = 'eta'
    boundary = 'top'
    value = 0
  [../]
  [./left_eta]
    type = DirichletBC
    variable = 'eta'
    boundary = 'left'
    value = 1
  [../]
  [./right_eta]
    type = DirichletBC
    variable = 'eta'
    boundary = 'right'
    value = 0
  [../]
  [./bottom_w]
    type = NeumannBC
    variable = 'w'
    boundary = 'bottom'
    value = 0
  [../]
  [./top_w]
    type = NeumannBC
    variable = 'w'
    boundary = 'top'
    value = 0.0
  [../]
  [./left_w]
    type = NeumannBC
    variable = 'w'
    boundary = 'left'
    value = 0
  [../]
  [./right_w]
    type = DirichletBC
    variable = 'w'
    boundary = 'right'
    value = 0.0
  [../]
  [./left_pot]
    type = DirichletBC
    variable = 'pot'
    boundary = 'left'
    value = -0.45
  [../]
  [./right_pot]
    type = DirichletBC
    variable = 'pot'
    boundary = 'right'
    value = 0
  [../]
[]

[Kernels]
  [./w_dot]
    # First part of equation 3 in main text . chi*dw/dt
    type = SusceptibilityTimeDerivative
    variable = w
    f_name = chi
    args = 'w'
  [../]
  [./Diffusion1]
    # Intrinsic diffusion part of equation 3 in main text.
    type = MatDiffusion
    variable = w
    diffusivity = D
  [../]
  [./Diffusion2]
    # Migration.
    type = Migration
    variable = w
    cv = eta
    Q_name = 0.
    QM_name = DN
    cp = pot
  [../]
  [./coupled_etadot]
    # Coupling between w and eta.
    type = CoupledSusceptibilityTimeDerivative
    variable = w
    v = eta
    f_name = ft
    args = 'eta'
  [../]
  [./Cond]
    # Conduction, left handside of equation 4 in main text.
    type = Conduction
    variable = pot
    cp = eta
    cv = w
    Q_name = Le1
    QM_name = 0.
  [../]
  [./coupled_pos]
    # Source term for Equation 4 in main text.
    type = CoupledSusceptibilityTimeDerivative
    variable = pot
    v = eta
    f_name = ft2
    args = 'eta'
  [../]
  [./BV]
    # Bulter-volmer equation, right hand side of Equation 1 in main text.
    type = Kinetics
    variable = eta
    f_name = G
    cp = pot
    cv = eta
  [../]
  [./AC_bulk]
    # Driving force from switching barrier, right hand side of Equation 1 in main text.
    type = AllenCahn
    variable = eta
    f_name = FF
  [../]
  [./AC_int]
    # interfacial energy
    type = ACInterface
    variable = eta
  [../]
  [./Noiseeta]
    type = LangevinNoise
    variable = eta
    amplitude = 0.04
  [../]
  [./e_dot]
    # deta/dt
    type = TimeDerivative
    variable = eta
  [../]
[]

[Materials]
  [./constants]
    # kappa_op: gradient coefficient;  M0:diffucion coefficient of Li+ in electrolyte
    #  S1, S2 conductivity of electrode and electrolyte; L: kinetic coefficient; Ls: electrochemical kinetic coefficient; B: Barrier height;
    #  es, el: difference in the chemical potential of lithium and neutral components on the electrode/electrolyte phase at initial equilibrium state;
    # us, ul: free energy density of the electrode/electrolyte phases. Defined in Ref. 20 and 26 of the main text; A: prefactor; AA: nF/(R*T);
    # dv is the ratio of site density for the electrode/electrolyte phases; ft2: normalized used in Equation 4.
    type = GenericConstantMaterial
    prop_names  = 'kappa_op  M0     S1    S2     L    Ls       B   es       el    A     ul    us    AA  dv   ft2'
    prop_values = '0.3   317.9   1000000 1.19   6.25   0.001  2.4  -13.8  2.631   1.0   0.0695 13.8   38.69 5.5 0.0074'
  [../]
  [./liquid_GrandPotential]
    # grand potential of electrolyte phase
    type = DerivativeParsedMaterial
    function = 'ul-A*log(1+exp((w-el)/A))'
    args = 'w'
    f_name = f1
    material_property_names = 'A ul el'
  [../]
  [./solid_GrandPotential]
    # grand potential of electrode phase
    type = DerivativeParsedMaterial
    function = 'us-A*log(1+exp((w-es)/A))'
    args = 'w'
    f_name = f2
    material_property_names = 'A us es'
  [../]
  [./switching_function]
    #interpolation function h
    type = SwitchingFunctionMaterial
    eta ='eta'
    h_order = HIGH
  [../]
  [./barrier_function]
    # Barrier function g
    type = BarrierFunctionMaterial
    eta = eta
  [../]
  [./total_GrandPotential]
    type = DerivativeTwoPhaseMaterial
    args = 'w'
    eta = eta
    fa_name = f1
    fb_name = f2
    derivative_order = 2
    W = 2.4
  [../]
  [./coupled_eta_function]
    # Coupling between eta and w
    type = DerivativeParsedMaterial
    function = '-(cs*dv-cl)*dh'  # in this code cs=-cs h=eta dh=1
    args = ' w eta'
    f_name = ft
    material_property_names = 'dh:=D[h,eta] h dv cs:=D[f2,w] cl:=D[f1,w]'
    derivative_order = 1
  [../]
  [./susceptibility]
    type = DerivativeParsedMaterial
    function = '-d2F1*(1-h)-d2F2*h*dv'
    args = 'w'
    f_name = chi
    derivative_order = 1
    material_property_names = 'h dv d2F1:=D[f1,w,w] d2F2:=D[f2,w,w]'
  [../]
  [./Mobility_coefficient]
    # Mobility defined by D*c/(R*T), whereR*T is normalized by the chemical potential
    # M0*(1-h) is the effective diffusion coefficient; cl*(1-h) is the ion concentration
    type = DerivativeParsedMaterial
    function = '-M0*(1-h)*cl*(1-h)'  #c is -c
    f_name = D
    args = 'eta w'
    derivative_order = 1
    material_property_names = ' M0 cl:=D[f1,w] h'
  [../]
  [./Free]
    # Energy of the barrier
    type = DerivativeParsedMaterial
    f_name = FF
    material_property_names = 'B'
    args='eta'
    function = 'B*eta*eta*(1-eta)*(1-eta)'
    derivative_order = 1
  [../]
  [./Migration_coefficient]
    # Migration coefficient.
    type = DerivativeParsedMaterial
    function = '-cl*(1-h)*AA*M0*(1-h)'
    args = 'eta w'
    f_name = DN
    derivative_order = 1
    material_property_names = 'M0 AA cl:=D[f1,w] h'
  [../]
  [./Bultervolmer]
    type = DerivativeParsedMaterial
    function = 'Ls*(exp(pot*AA/2.)+14.89*cl*(1-h)*exp(-pot*AA/2.))*dh'
    args = 'pot eta w'
    f_name = G
    derivative_order = 1
    material_property_names = 'Ls dh:=D[h,eta] h cl:=D[f1,w] AA'
    outputs = exodus
  [../]
  [./concentration]
    # output the ion concentration
    type = ParsedMaterial
    f_name = c
    args='eta w'
    material_property_names = 'h dFl:=D[f1,w]'
    function = '-dFl*(1-h)'
    outputs = exodus
  [../]
  [./Le1]
    # Effective conductivity
    type = DerivativeParsedMaterial
    f_name = Le1
    args = 'eta'
    material_property_names = 'S1 S2 h'
    function = 'S1*h+S2*(1-h)'
    derivative_order = 1
  [../]
[]

[GlobalParams]
  enable_jit = false           # Disable JIT
[]

[Preconditioning]
  [./SMP]
    type = SMP
    full = true
  [../]
[]

[Executioner]
  type = Transient
  scheme = 'bdf2'
  solve_type = 'PJFNK'
  l_max_its = 50
  l_tol = 1e-4
  nl_max_its = 20
  nl_rel_tol = 1e-6
  nl_abs_tol = 1e-6
  dt = 0.02
  end_time = 400
  petsc_options_iname = '-pc_type -ksp_grmres_restart -sub_ksp_type
                         -sub_pc_type -pc_asm_overlap'
  petsc_options_value = 'asm      121                  preonly
                         lu          8'
[]

[Outputs]
  exodus = true
  csv = true
  execute_on = 'TIMESTEP_END'
  [./other]        # creates input_other.e output every 30 timestep
     type = Exodus
     interval = 30
  [../]
[]

```

**Reply by GiudGiud** [reply]
where did you get that model from?
I d first have a look in the relevant module to try to see if there s an updated syntax for this input

**Reply by wowodejiajia** [reply]
The article is here.

Phase-Field Simulations of Lithium Dendrite Growth with Open-Source Software
Zijian Hong and Venkatasubramanian Viswanathan
ACS Energy Letters 2018 3 (7), 1737-1743
DOI: 10.1021/acsenergylett.8b01009

**Comment by wowodejiajia** [comment]
The article is here : https://pubs.acs.org/doi/10.1021/acsenergylett.8b01009

**Reply by GiudGiud** [reply]
added the link. 2018 so the syntax could have evolved since then. 
Let s have @laagesen @SudiptaBiswas  or @amjokisaari pitch in since this is phase field

**Reply by wowodejiajia** [reply]
Thank you very much.

**Comment by wowodejiajia** [comment]
I have changed solve_type =Newton  in [Executioner]. And then I got this result.

The following total 5 aux variables:
  G
  c
  dG/deta
  dG/dpot
  dG/dw
are added for automatic output by MaterialOutputAction.
Finished Setting Up                                                                      [  0.62 s] [  101 MB]
Framework Information:
MOOSE Version:           git commit b0f5d8a7ab on 2021-10-25
LibMesh Version:
PETSc Version:           3.15.1
SLEPc Version:           3.15.1
Current Time:            Fri Oct 29 11:06:25 2021
Executable Timestamp:    Wed Oct 27 16:50:13 2021

Parallelism:
  Num Processors:          2
  Num Threads:             2

Mesh:
  Parallel Type:           replicated
  Mesh Dimension:          2
  Spatial Dimension:       2
  Nodes:
    Total:                 40401
    Local:                 20326
    Min/Max/Avg:           20075/20326/20200
  Elems:
    Total:                 40000
    Local:                 20000
    Min/Max/Avg:           20000/20000/20000
  Num Subdomains:          1
  Num Partitions:          2
  Partitioner:             metis

Nonlinear System:
  Num DOFs:                121203
  Num Local DOFs:          60978
  Variables:               { "w" "eta" "pot" }
  Finite Element Types:    "LAGRANGE"
  Approximation Orders:    "FIRST"

Auxiliary System:
  Num DOFs:                200000
  Num Local DOFs:          100000
  Variables:               { "G" "c" "dG/deta" "dG/dpot" "dG/dw" }
  Finite Element Types:    "MONOMIAL"
  Approximation Orders:    "CONSTANT"

Execution Information:
  Executioner:             Transient
  TimeStepper:             ConstantDT
  Solver Mode:             NEWTON
  MOOSE Preconditioner:    SMP
......
Time Step 10, time = 0.00500004, dt = 2e-14
 0 Nonlinear |R| = 3.032890e+05
      0 Linear |R| = 3.032890e+05
      1 Linear |R| = 6.471922e+00
......
10 Linear |R| = 1.716624e-03
     11 Linear |R| = 3.090150e-04
     12 Linear |R| = 1.064597e-04
20 Nonlinear |R| = 1.630292e+00
Nonlinear solve did not converge due to DIVERGED_MAX_IT iterations 20
 Solve Did NOT Converge!
  Finished Solving                                                                       [ 53.39 s] [    0 MB]
Aborting as solve did not converge


*** ERROR ***
The following error occurred in the object "TimeStepper", of type "ConstantDT".

Solve failed and timestep already at or below dtmin, cannot continue!



*** ERROR ***
The following error occurred in the object "TimeStepper", of type "ConstantDT".

Solve failed and timestep already at or below dtmin, cannot continue!

application called MPI_Abort(MPI_COMM_WORLD, 1) - process 1
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
Finished: Exit code: 1

What should I do?

**Reply by GiudGiud** [reply]
so the non linear residual went from 3e5 to 1.6 so the good news is that it looks like your simulation is converging. It just needs more iterations. 

So in the [Executioner] block, please specify:
nl_max_its = 100
to let it take more non linear iterations. 

You can also lower the tolerances if the residual is low enough : nl_abs_tol and nl_rel_tol can be used for that. 

See the list of those solver parameters under Solver Parameters (at the bottom of the page)
https://mooseframework.inl.gov/source/executioners/Transient.html

**Reply by dschwen** [reply]
Is your initial condition even close to a solution at time zero? If not then it is not unusual for the first time step to require a _lot_ of iterations to solve.

**Reply by wowodejiajia** [reply]
Thank you for your answers. But it doesn't work. I realized it was parallelism that was the problem. Programs can converge without running in parallel. Can you explain why that is?

Looking forward to answer.

**Reply by dschwen** [reply]
not every preconditioner works equally well in parallel. you could try a simple LU (which in parallel would use SuperLU automatically)

**Comment by wowodejiajia** [comment, marked as answer]
All the documents are here.

[Conduction-C.txt](https://github.com/idaholab/moose/files/7472224/Conduction-C.txt)
[Conduction-h.txt](https://github.com/idaholab/moose/files/7472225/Conduction-h.txt)
[ex01-P.txt](https://github.com/idaholab/moose/files/7472226/ex01-P.txt)
[Kinetics-C.txt](https://github.com/idaholab/moose/files/7472227/Kinetics-C.txt)
[Kinetics-h.txt](https://github.com/idaholab/moose/files/7472228/Kinetics-h.txt)
[Migration-C.txt](https://github.com/idaholab/moose/files/7472229/Migration-C.txt)
[Migration-h.txt](https://github.com/idaholab/moose/files/7472230/Migration-h.txt)

**Reply by dschwen** [reply]
Converges fine for me with `-pc_type lu`. Step three struggles with a dt of 0.02 but converges after automatically cutting to 0.01. Then it continues to converge fine.

**Reply by dschwen** [reply]
You should probably look into adaptive timestepping as during later times the simulation converges more quickly.

```
Time Step 16, time = 0.3, dt = 0.02
 0 Nonlinear |R| = 3.974561e+01
      0 Linear |R| = 3.974561e+01
      1 Linear |R| = 1.439269e+00
      2 Linear |R| = 3.138209e-01
      3 Linear |R| = 1.718015e-02
      4 Linear |R| = 1.651372e-03
 1 Nonlinear |R| = 2.183084e+01
      0 Linear |R| = 2.183084e+01
      1 Linear |R| = 3.271910e-02
      2 Linear |R| = 1.750573e-03
 2 Nonlinear |R| = 9.665931e-01
      0 Linear |R| = 9.665931e-01
      1 Linear |R| = 2.149134e-03
      2 Linear |R| = 1.422680e-04
      3 Linear |R| = 8.701771e-06
 3 Nonlinear |R| = 3.019655e-03
      0 Linear |R| = 3.019655e-03
      1 Linear |R| = 3.152191e-05
      2 Linear |R| = 8.527107e-07
      3 Linear |R| = 1.259573e-07
 4 Nonlinear |R| = 3.128464e-05
 Solve Converged!
  Finished Solving                                                                       [  8.26 s] [    0 MB]
```

(that's on 8 cores)

**Reply by dschwen** [reply]
and

```
Time Step 30, time = 0.58, dt = 0.02
0 Nonlinear |R| = 3.545873e+01
      0 Linear |R| = 3.545873e+01
      1 Linear |R| = 8.017880e-01
      2 Linear |R| = 6.858638e-02
      3 Linear |R| = 5.232085e-03
      4 Linear |R| = 4.528376e-04
 1 Nonlinear |R| = 7.029588e-01
      0 Linear |R| = 7.029588e-01
      1 Linear |R| = 4.787240e-03
      2 Linear |R| = 3.551958e-04
      3 Linear |R| = 4.642392e-05
 2 Nonlinear |R| = 1.644062e-03
      0 Linear |R| = 1.644062e-03
      1 Linear |R| = 1.856705e-05
      2 Linear |R| = 2.448408e-06
      3 Linear |R| = 2.262082e-07
      4 Linear |R| = 1.217206e-08
 3 Nonlinear |R| = 1.516930e-05
 Solve Converged!
  Finished Solving                                                                       [  6.79 s] [    0 MB]
```

**Reply by dschwen** [reply]
Here bd240d4394ce5d1 is the slightly updated version of your files that I ran to get above mentioned convergence.

**Reply by wowodejiajia** [reply]
Thank you very much for your answer. I have one more question.
You used eight cores. Which cores are used for MPI and which cores are used for Threads?

**Reply by dschwen** [reply]
By default no threading is used, so I used all cored for MPI ranks.

**Reply by wowodejiajia** [reply]
Thank you.
